libs/mpmc_queue/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the number of elements in the queue.
+    pub fn len(&self) -> usize {
+        self.len.load(Ordering::Acquire)
+    }
+
+    /// Appends an item to the queue.
+    pub fn push(&self, item: T) {
+        let node = box_pointer(Node::new(item));
+        unsafe { self.push_inner(node, node, 1) };
+    }
+
+    /// Appends several items to the queue.
+    pub fn push_batch<I>(&self, mut iterator: I)
+    where
+        I: Iterator<Item = T>,
+    {
+        let first = match iterator.next() {
+            Some(item) => box_pointer(Node::new(item)),
+            None => return,
+        };
+
+        let mut tail = first;
+        let mut len = 1;
+
+        for next in iterator {
+            let next = box_pointer(Node::new(next));
+            // SAEFTY: The only other pointer to the tail is stored in the second-to-last
+            // node. We own first, hence we own all the nodes, hence that reference is not
+            // being used.
+            unsafe { tail.as_mut() }.next = Some(next);
+            tail = next;
+            len += 1;
+        }
+
+        unsafe { self.push_inner(first, tail, len) };
+    }
+
+    /// Appends a batch of nodes to the queue.
+    ///
+    /// # Safety
+    ///
+    /// `head` must be the start of the batch, and `tail` must point to the end
+    /// of the batch. The batch must be `len` nodes long.
+    unsafe fn push_inner(&self, head: NonNull<Node<T>>, tail: NonNull<Node<T>>, len: usize) {
+        let mut pointers = self.pointers.lock();
+
+        if let Some(mut tail_pointer) = pointers.tail {
+            // SAEFTY: The only other pointer to the tail is stored in the second-to-last
+            // node. We hold the lock to pointers, hence we own all the nodes, hence that
+            // reference is not being used.
+            let tail = unsafe { tail_pointer.as_mut() };
+            tail.next = Some(head);
+        } else {
+            debug_assert!(pointers.head.is_none());
+            pointers.head = Some(head);
+        }
+        pointers.tail = Some(tail);
+
+        self.len.fetch_add(len, Ordering::Release);
+    }
+
+    /// Pops a node from the front of the queue.
+    pub fn pop(&self) -> Option<T> {
+        if self.is_empty() {
+            return None;
+        }
+
+        let mut pointers = self.pointers.lock();
+
+        // This will return none if another thread popped the last task between us
+        // checking the fast path and now.
+        let current_head = pointers.head.take()?;
+        // SAFETY: current_head is a valid pointer.
+        pointers.head = unsafe { current_head.as_ref() }.next;
+
+        // If we are the last node in the list:
+        if pointers.head.is_none() {
+            pointers.tail = None;
+        }
+
+        self.len.fetch_sub(1, Ordering::Release);
+        // SAFETY: current_head is a valid pointer.
```

Comment:
i think `Box::from_raw()` also requires that the pointer was obtained from a `Box` originally, too. Obviously you ensure that via usage of `box_pointer()`, but I still think it merits an explicit comment here.

Reply:
specific details here: https://doc.rust-lang.org/std/boxed/index.html#memory-layout

so it's not that it had to come from a `Box`, but the fact that it came from a `Box` does satisfy the `Layout`-related invariants.